### ISSUE TITLE
VUID-RuntimeSpirv-cooperativeMatrixFlexibleDimensions-10166 false positive when a workgroup-scope cooperative matrix uses LocalSizeId with a specialization constant

### ISSUE THREAD

**Issue by Indie-Siggi** [KhronosGroup/Vulkan-ValidationLayers#12363]
**Environment:**
 - OS: Linux (kernel 6.19.14) — but the bug is in the layer's static analysis and is OS/driver-independent
 - GPU and driver version: AMD RX 9070 XT, RADV / Mesa (`GFX1201`). The defect is in the layer's static analysis (driver-independent) — it triggers for any workgroup-scope cooperative-matrix shader whose local size comes from a specialization constant, on any implementation that advertises such properties.
 - SDK or header version: Validation Layers **1.4.350.0** (Vulkan SDK 1.4.350); also confirmed on **`main`** @ HEAD `633389498` (2026-06-05). Line numbers below are from `main`.
 - Options enabled: default core validation

**Describe the Issue**

`vkCreateComputePipelines` reports **VUID-RuntimeSpirv-cooperativeMatrixFlexibleDimensions-10166** for a
workgroup-scope cooperative-matrix shader whose local workgroup size is supplied through a **specialization
constant** via `LocalSizeId`, even when the specialized size exactly matches the advertised
`workgroupInvocations`. It is a false positive: the layer cannot resolve the specialized local size and treats
the workgroup size as `1`.

Root cause — the workgroup-scope match requires the shader's total workgroup size to equal the property's
`workgroupInvocations` (`cc_spirv.cpp:639`, and the fixed-size path at `:505`):

```cpp
valid &= property.scope != VK_SCOPE_WORKGROUP_KHR || workgroup_size == property.workgroupInvocations;
```

`workgroup_size` (`cc_spirv.cpp:295`) comes from `Module::FindLocalSize()`, which for `LocalSizeId` resolves
each dimension through `Module::GetConstantValueById()` (`shader_module.cpp:1651-1653`). `FindLocalSize()` even
documents the limitation in the comment directly above it (`shader_module.cpp:1631`): *"Assumes shader module
contains no spec constants used to set the local size values."* And `GetConstantValueById()` returns **`1`** for
anything that is not a plain `OpConstant` — which includes specialization constants
(`shader_module.cpp:1711`):

```cpp
uint32_t Module::GetConstantValueById(uint32_t id) const {
    const Instruction* value = GetAnyConstantDef(id);
    // If this hit, most likley a runtime array (probably from VK_EXT_descriptor_indexing)
    // or unhandled specialization constants
    if (!value || value->Opcode() != spv::OpConstant) {
        // TODO - still not fixed
        // https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/6293
        return 1;
    }
    return value->GetConstantValue();
}
```

So when `local_size_x` is a spec constant, `workgroup_size` becomes `1 × 1 × 1 = 1`, the check becomes
`1 == workgroupInvocations` (e.g. `1 == 256`), and **every** workgroup-scope cooperative matrix whose size is
spec-constant-driven is rejected — regardless of the value actually supplied in `VkSpecializationInfo` at
pipeline creation. This is the same unresolved-spec-constant gap as #6293, surfacing as a coop-matrix VUID.

**Expected behavior**

No validation error: when the specialization constants supplied at pipeline creation make the real workgroup
size equal `workgroupInvocations` (and the matrix dimensions are valid multiples within range), the shader
should validate cleanly.

**Valid Usage ID**

`VUID-RuntimeSpirv-cooperativeMatrixFlexibleDimensions-10166`

**Additional context**

Steps to reproduce with `jeffbolznv/vk_cooperative_matrix_perf`, against a driver that advertises
workgroup-scope flexible-dimension cooperative matrices (e.g. a `VK_NV_cooperative_matrix2` implementation):

```sh
VK_LOADER_LAYERS_ENABLE=VK_LAYER_KHRONOS_validation ./vk_cooperative_matrix_perf --correctness
```

The `workgroup*` shaders declare `layout(local_size_x_id = 18, local_size_y = 1, local_size_z = 1) in;` and set
spec-constant id 18 to 128 or 256 at pipeline creation; the layer emits VUID-10166 for each workgroup-scope
flexible-dimension matrix.

Minimal standalone conditions:
1. A compute shader using `OpExecutionModeId ... LocalSizeId` where the x dimension id is an `OpSpecConstant`.
2. A workgroup-scope cooperative matrix (flexible dimensions) used in `OpCooperativeMatrixMulAddKHR`.
3. Create the pipeline with `VkSpecializationInfo` setting that spec constant equal to `workgroupInvocations`.

This is not a driver problem: with the layer loaded and reporting these errors, the dispatches still execute and
produce correct results — on RADV (`GFX1201`) the same `--correctness` run passes all 3356 configurations with
validation enabled — so the matrices are genuinely supported and correctly dispatched; only the static check is
wrong.

Suggested fix (at `cc_spirv.cpp:295`, used by the checks at `:505` and `:639`):
1. Resolve the pipeline's `VkSpecializationInfo` before computing the local size, so `FindLocalSize()` /
   `GetConstantValueById()` return the real specialized values (also addresses the broader #6293 impact); or
2. Narrower: when a `LocalSizeId` dimension is an unresolved specialization constant (i.e. `GetConstantValueById()`
   falls through to the `1` placeholder), skip the `workgroup_size == workgroupInvocations` equality rather than
   silently using `1`.

Related: #6293 (the underlying "unhandled specialization constants → returns 1" gap).

In the output below, the requested workgroup-scope `e4m3 × e4m3 → fp32` matrix matches advertised
flexible-dimension property `[1]` on type, scope, and granularity (the dimensions are multiples of 16) — the
only failing condition is the workgroup-size equality, where the layer's mis-computed `1` is compared against
the real `workgroupInvocations = 256`.

<details>
<summary>validation output (verbatim; long property lists elided as noted)</summary>

```sh
Validation Error: [ VUID-RuntimeSpirv-cooperativeMatrixFlexibleDimensions-10166 ] | MessageID = 0x93d7111f
vkCreateComputePipelines(): pCreateInfos[0].stage shader [VK_SHADER_STAGE_COMPUTE_BIT] instruction
%151 = OpCooperativeMatrixMulAddKHR %62 %148 %149 %150
doesn't match a supported matrix VkCooperativeMatrixPropertiesKHR or VkPhysicalDeviceCooperativeMatrix2PropertiesNV
rows: 128, cols: 64, scope: VK_SCOPE_WORKGROUP_KHR, type: VK_COMPONENT_TYPE_FLOAT8_E4M3_EXT, use: 0
rows: 64, cols: 128, scope: VK_SCOPE_WORKGROUP_KHR, type: VK_COMPONENT_TYPE_FLOAT8_E4M3_EXT, use: 1
rows: 128, cols: 128, scope: VK_SCOPE_WORKGROUP_KHR, type: VK_COMPONENT_TYPE_FLOAT32_KHR, use: 2
rows: 128, cols: 128, scope: VK_SCOPE_WORKGROUP_KHR, type: VK_COMPONENT_TYPE_FLOAT32_KHR, use: 2
# ... fixed-size VkCooperativeMatrixPropertiesKHR list (all VK_SCOPE_SUBGROUP_KHR) elided ...
# advertised flexible-dimension properties (VkPhysicalDeviceCooperativeMatrix2PropertiesNV), excerpt:
[1] MGranularity = 16 | NGranularity = 16 | KGranularity = 16 | AType = VK_COMPONENT_TYPE_FLOAT8_E4M3_EXT | BType = VK_COMPONENT_TYPE_FLOAT8_E4M3_EXT | CType = VK_COMPONENT_TYPE_FLOAT32_KHR | ResultType = VK_COMPONENT_TYPE_FLOAT32_KHR | scope = VK_SCOPE_WORKGROUP_KHR | workgroupInvocations = 256
# ... remaining flexible properties elided ...
```

</details>

<sub>Disclosure: this analysis was AI-assisted (Claude Code) and reviewed by me against the cited source on
`main` (HEAD `633389498`) and tag `vulkan-sdk-1.4.350.0`.</sub>

**Comment by spencer-lunarg** [comment]
@Indie-Siggi thanks for this, but all this issue information is just a lot of noise (and some just seems to be off the rails TBH) and will take longer for anyone to fix. Please provide

1. The error message (which you did!)
2. A SPIR-V file reproduce
3. The spec constant values you set at pipeline/shaderObject creation time

**Comment by Indie-Siggi** [comment]
Thank you for the reply, I am still learning and feedback is welcome. Here are the three items, plus the device
property the check needs in order to fire, so this is reproducible without my hardware.

### 1. Error message

`vkCreateComputePipelines` on the attached shader reports this (the full block was in the first post):

```
Validation Error: [ VUID-RuntimeSpirv-cooperativeMatrixFlexibleDimensions-10166 ]
vkCreateComputePipelines(): pCreateInfos[0].stage shader [VK_SHADER_STAGE_COMPUTE_BIT] instruction
%151 = OpCooperativeMatrixMulAddKHR %62 %148 %149 %150
doesn't match a supported matrix VkCooperativeMatrixPropertiesKHR or VkPhysicalDeviceCooperativeMatrix2PropertiesNV
rows: 128, cols: 64,  scope: VK_SCOPE_WORKGROUP_KHR, type: VK_COMPONENT_TYPE_FLOAT8_E4M3_EXT, use: 0
rows: 64,  cols: 128, scope: VK_SCOPE_WORKGROUP_KHR, type: VK_COMPONENT_TYPE_FLOAT8_E4M3_EXT, use: 1
rows: 128, cols: 128, scope: VK_SCOPE_WORKGROUP_KHR, type: VK_COMPONENT_TYPE_FLOAT32_KHR,     use: 2
```

The same pipeline also raises the sibling check VUID-RuntimeSpirv-cooperativeMatrixFlexibleDimensions-10165
(on the `OpTypeCooperativeMatrixKHR` itself). Both come from the same workgroup-size evaluation, so one fix
clears both.

### 2. SPIR-V file

Attached: `workgroupe4m3_fp32.spv` (SPIR-V 1.6). It is the `workgroup.comp` shader from github
`jeffbolznv/vk_cooperative_matrix_perf`, built with:

```sh
glslangValidator --target-env spirv1.6 -DA_TYPE=floate4m3_t -DC_TYPE=float32_t -V workgroup.comp -o workgroupe4m3_fp32.spv
```

The relevant part is that the local workgroup size comes from a specialization constant (SpecId 18), with a
default of `1` in the module:

```
OpExecutionModeId %main LocalSizeId %7 %uint_1 %uint_1   ; local_size = (%7, 1, 1)
OpDecorate        %7 SpecId 18                            ; %7 is spec constant 18
%7 = OpSpecConstant %uint 1                              ; default value 1
```

[workgroupe4m3_fp32.spv.zip](https://github.com/user-attachments/files/28664775/workgroupe4m3_fp32.spv.zip)


### 3. Spec-constant values at pipeline creation (`VkSpecializationInfo`)

- **SpecId 18 = 256** is `local_size_x`, so the real workgroup size is `256 × 1 × 1 = 256`.
- SpecId 3 (`TILE_M`) = 128, SpecId 4 (`TILE_N`) = 128, SpecId 5 (`TILE_K`) = 64 give the 128×128 coopmat in
  the error.

The shader declares about 21 constants. The rest (strides, `M`, `N`) do not affect this VUID. The full set is
the `specData[]` array in `src/vk_cooperative_matrix_perf.cpp` from `jeffbolznv/vk_cooperative_matrix_perf`.

### Device precondition (needed for the check to run)

These VUIDs only execute when `VkPhysicalDeviceCooperativeMatrix2FeaturesNV::cooperativeMatrixFlexibleDimensions`
is enabled and the device advertises a matching workgroup-scope flexible-dimension property. To reproduce, the
device (or a profile/MockICD) must advertise this `VkCooperativeMatrixFlexibleDimensionsPropertiesNV`:

```
MGranularity = 16, NGranularity = 16, KGranularity = 16
AType = FLOAT8_E4M3, BType = FLOAT8_E4M3, CType = FLOAT32, ResultType = FLOAT32
scope = VK_SCOPE_WORKGROUP_KHR, workgroupInvocations = 256
```

On real hardware that is any RDNA4 GPU with Mesa RADV's coopmat2 enabled, or an NVIDIA
`VK_NV_cooperative_matrix2` implementation.

### Why it's a false positive

With `SpecId 18 = 256` the real workgroup size equals `workgroupInvocations = 256`, and the matrix matches the
property above on type, scope, and granularity (128 and 64 are multiples of 16), so it should validate, but the
layer rejects it. With validation enabled the pipeline is still created and dispatched, and on RDNA4
(`GFX1201`, RADV coopmat2) the tool's `--correctness` run passes 3356/3356 (0 failures) while still printing
these VUIDs. So the matrices are genuinely supported, and only the static check rejects them.

<details>
<summary>Where the workgroup size goes wrong (source pointers, 1.4.350.0)</summary>

The matched flexible property is rejected only by `workgroup_size == property.workgroupInvocations`
(`cc_spirv.cpp:638` for 10166, and the skip at `:504` for 10165). `workgroup_size` is
`local_size.x * local_size.y * local_size.z` (`cc_spirv.cpp:294`), where `local_size` comes from
`Module::FindLocalSize()`, which resolves the `LocalSizeId` dimension via `GetConstantValueById()`
(`shader_module.cpp:1650`). For a specialization constant that function returns `1` (its comment notes it does
not handle spec constants, at `shader_module.cpp:1630` and `:1704`), so `workgroup_size` is computed as `1` and
the comparison becomes `1 == 256`. The matrix dimensions, in contrast, are resolved with the specialized
values, which is why the error prints them correctly as 128/64/128. The `return 1` branch in
`GetConstantValueById()` carries a TODO that points at #6293. The full breakdown is in the original post.

</details>

*Disclosure: this follow-up was prepared with AI assistance (Claude Code) and reviewed by me. The values were
verified on RDNA4 hardware (RADV coopmat2) and against VVL source at tag `vulkan-sdk-1.4.350.0`.*

**Comment by spencer-lunarg** [comment]
ok, thanks this is more useful, I will look into it when I get a chance before the next SDK release in July

If it is actually NOT a false positive, then the "real bug" is the fact the error message is not clear enough as I can't tell neither from the current message (it is hard because there are so many properties for Coopmat)

**Comment by jeffbolznv** [comment]
My fix was bogus. @Indie-Siggi can you share the complete set of supported cooperative matrix properties? I wonder if the issue is that the app doesn't check that the workgroup size is supported, and this is a legitimate app bug the layers are catching.

**Comment by Indie-Siggi** [comment]
First I would like to say that I made an error with the false positive assumption, this was a chained misinterpretation of my AI-Assistant. Which even with the guide to reproduce the error didn't lead to the right conclusions. So sorry for wasting your time on the false lead. This is not a bug from Vulkan-ValidationLayers.

Furthermore you're right, and thanks for the hint. It seems to be an app bug in `vk_cooperative_matrix_perf`, not a VVL false positive. My earlier comment was wrong and I want to correct the record.

**Complete supported set (RX 9070 XT, RADV GFX1201, Mesa coopmat2):** 20 fixed `VkCooperativeMatrixPropertiesKHR`,
all `SUBGROUP`, all 16x16x16. 40 flexible `VkCooperativeMatrixFlexibleDimensionsPropertiesNV`, granularity 16,
in two scopes:

```
SUBGROUP  -> workgroupInvocations = 0     (20 entries)
WORKGROUP -> workgroupInvocations = 256   (20 entries)   // every workgroup-scope prop needs 256; there is no 128
```

<details>
<summary>Complete enumerated set (60 entries). <code>saturatingAccumulation</code> is from RADV source, not the dump (which omits that field); true only for the signed-int saturating variants. The requested list is also in the attached file.</summary>

```
== VkCooperativeMatrixPropertiesKHR (20, all subgroup, M=N=K=16) ==
  A     B     C     Result  scope     satAccum
  e4m3  e4m3  fp32  fp32    subgroup  false
  e4m3  e5m2  fp32  fp32    subgroup  false
  e5m2  e4m3  fp32  fp32    subgroup  false
  e5m2  e5m2  fp32  fp32    subgroup  false
  u8    u8    u32   u32     subgroup  false
  u8    u8    s32   s32     subgroup  false
  u8    u8    s32   s32     subgroup  true
  u8    s8    u32   u32     subgroup  false
  u8    s8    s32   s32     subgroup  false
  u8    s8    s32   s32     subgroup  true
  s8    u8    u32   u32     subgroup  false
  s8    u8    s32   s32     subgroup  false
  s8    u8    s32   s32     subgroup  true
  s8    s8    u32   u32     subgroup  false
  s8    s8    s32   s32     subgroup  false
  s8    s8    s32   s32     subgroup  true
  fp16  fp16  fp16  fp16    subgroup  false
  bf16  bf16  bf16  bf16    subgroup  false
  fp16  fp16  fp32  fp32    subgroup  false
  bf16  bf16  fp32  fp32    subgroup  false

== VkCooperativeMatrixFlexibleDimensionsPropertiesNV (40, granularity 16) ==
  A     B     C     Result  scope     satAccum wgInvoc
  e4m3  e4m3  fp32  fp32    subgroup  false    0
  e4m3  e4m3  fp32  fp32    workgroup false    256
  e4m3  e5m2  fp32  fp32    subgroup  false    0
  e4m3  e5m2  fp32  fp32    workgroup false    256
  e5m2  e4m3  fp32  fp32    subgroup  false    0
  e5m2  e4m3  fp32  fp32    workgroup false    256
  e5m2  e5m2  fp32  fp32    subgroup  false    0
  e5m2  e5m2  fp32  fp32    workgroup false    256
  u8    u8    u32   u32     subgroup  false    0
  u8    u8    u32   u32     workgroup false    256
  u8    u8    s32   s32     subgroup  false    0
  u8    u8    s32   s32     workgroup false    256
  u8    u8    s32   s32     subgroup  true     0
  u8    u8    s32   s32     workgroup true     256
  u8    s8    u32   u32     subgroup  false    0
  u8    s8    u32   u32     workgroup false    256
  u8    s8    s32   s32     subgroup  false    0
  u8    s8    s32   s32     workgroup false    256
  u8    s8    s32   s32     subgroup  true     0
  u8    s8    s32   s32     workgroup true     256
  s8    u8    u32   u32     subgroup  false    0
  s8    u8    u32   u32     workgroup false    256
  s8    u8    s32   s32     subgroup  false    0
  s8    u8    s32   s32     workgroup false    256
  s8    u8    s32   s32     subgroup  true     0
  s8    u8    s32   s32     workgroup true     256
  s8    s8    u32   u32     subgroup  false    0
  s8    s8    u32   u32     workgroup false    256
  s8    s8    s32   s32     subgroup  false    0
  s8    s8    s32   s32     workgroup false    256
  s8    s8    s32   s32     subgroup  true     0
  s8    s8    s32   s32     workgroup true     256
  fp16  fp16  fp16  fp16    subgroup  false    0
  fp16  fp16  fp16  fp16    workgroup false    256
  bf16  bf16  bf16  bf16    subgroup  false    0
  bf16  bf16  bf16  bf16    workgroup false    256
  fp16  fp16  fp32  fp32    subgroup  false    0
  fp16  fp16  fp32  fp32    workgroup false    256
  bf16  bf16  fp32  fp32    subgroup  false    0
  bf16  bf16  fp32  fp32    workgroup false    256
```

</details>

[2026-10-radv-coopmat-properties-compact.txt](https://github.com/user-attachments/files/28807155/2026-10-radv-coopmat-properties-compact.txt)

**Why it's seemingly an app bug.** The workgroup-scope test loop hardcodes the launch size and never checks it against the
property's `workgroupInvocations` (verbatim, `src/vk_cooperative_matrix_perf.cpp` @ master `d2c4ac1`):

```cpp
1055:  for (unsigned int workgroupSize = 32; workgroupSize <= 256; workgroupSize *= 2) {
...
1121:      switch (tt) {
1122:      case TT_WORKGROUP:
1123:          if (workgroupSize != 128 && workgroupSize != 256) {
1124:              continue;
1125:          }
1126:          break;
```

That `workgroupSize` is written straight into the `local_size_x` spec constant (id 18):

```cpp
1287:          workgroupSize, // invocations per workgroup
```

On RADV the only workgroup-scope property requires `workgroupInvocations = 256`, so the `workgroupSize = 128`
pipelines request a matrix no property supports, and the layer is right to reject them.

**VVL resolves the specialized size correctly.** I re-ran `--correctness` with `duplicate_message_limit = 0` and
a marker before each `vkCreateComputePipelines`, then attributed every VUID to its pipeline:

```
WORKGROUP wg=128 : 68 pipelines -> 136x VUID-10166, 408x VUID-10165
WORKGROUP wg=256 : 68 pipelines ->   0,              0
subgroup         : 3219 pipelines -> 0,              0
```

The `wg=256` pipelines (size == workgroupInvocations) raise nothing. If the layer computed `workgroup_size = 1`
(what I claimed earlier), `256` would also fail `1 == 256`. It doesn't: VVL freezes the spec constants
(`CreateFreezeSpecConstantValuePass`, then rebuilds the module in `cc_spirv.cpp`) before the coopmat check, so
`FindLocalSize` reads the real size.

**What I got wrong, and why.** My original report was wrong because I misread a *deduplicated* validation log.
Two things about that log hid the truth: the VUID-10166 message names the **matrix shape** (e.g. `128x64`) but
**not the workgroup launch size**, so a 128-thread launch (`wg=128`) and a 256-thread launch (`wg=256`) of the
same matrix print identical text; and the layer's duplicate-message suppression was on, so the 136 actual errors
collapsed to ~20 distinct lines. Together that meant I could not tell the 128-thread launches from the 256-thread
ones, because what distinguishes them is not in the message. I saw `128x64` in the rejected set and assumed both
launch sizes failed. Only the 128-thread ones did. Re-running with suppression off (`duplicate_message_limit = 0`)
and a marker printing the workgroup size before each `vkCreateComputePipelines` made it unambiguous: all 136
errors came from `wg=128`, none from `wg=256`.

Two things compounded it. My "`GetConstantValueById` returns 1" explanation overlooked that the layer freezes the
specialization constants before this check, so it does resolve the real workgroup size (the `128x128` dimensions
in the error are themselves the frozen `TILE_*` constants, which already showed the module was specialized). And
the reproduction steps I posted told you to set the spec constant **equal to** `workgroupInvocations` (256), which
is exactly the case that validates cleanly. Sorry for the noise.

**Suggested fix (app side):** gate the workgroup-scope `workgroupSize` on the matching property's `workgroupInvocations`
instead of hardcoding `{128, 256}` (capture `workgroupInvocations` in the property loop, then
`case TT_WORKGROUP: if (workgroupSize != workgroupInvocations) continue;`). I have it verified locally: on RADV
VUID-10166/10165 go 544 -> 0 and `--correctness` passes 3288/3288 (the 68 removed runs are exactly the invalid
`wg=128` launches).  This doesn't really belong here, but to get a full picture I placed the possible fix in this report.

Side note, not a VVL matter: on RADV the `wg=128` dispatches still pass `--correctness`, so RADV can run
128-invocation workgroup matrices but only advertises the 256 variant. Whether it should also advertise 128 is an
!41775 question; per spec the app must match an advertised property regardless, so the app fix stands on its own.

<sub>Disclosure: AI-assisted (Claude Code), reviewed by me. Verified on RDNA4 (RADV GFX1201, coopmat2 via Mesa
MR !41775) and against VVL source at `vulkan-sdk-1.4.350.0`.</sub>

**Comment by spencer-lunarg** [comment]
> Furthermore you're right, and thanks for the hint. It seems to be an app bug in vk_cooperative_matrix_perf, not a VVL false positive. 

Even if that is the case, I find it a "VVL bug" when from the error message itself, I can't deduce if it is a false positive or not... the VVL error message should be improved still
